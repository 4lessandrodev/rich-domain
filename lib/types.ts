/**
 * @summary The result is used to returns a operation result instead the own value.
 * @interface IResult<T, D, M>;
 * @classdesc on `T` refer to type of the value and `D` type of the error and `M` metaData type.
 * @default D is string.
 * @default M is empty object {}.
 * 
 * @method `value()` get result value. return null if result is failure.
 * @method `error()` get result error. returns null if result is success.
 * @method `isFail()` check is result is failure
 * @method `isOk()` check if result is success
 * @method `metaData()` get result metadata
 * @method `toObject()` get an object with result state
 * @method `execute()` execute a hook as command on fail or on success
 */
export interface IResult<T, D = string, M = {}> {
	value(): T;
	error(): D;
	isFail(): boolean;
	isOk(): boolean;
	metaData(): M;
	toObject(): IResultObject<T, D, M>;
	execute:<X, Y>(command: ICommand<X|void, Y>)=>IResultExecute<X, Y>;
}

/**
 * 
 */
export interface UID<T = string> {
	toShort(): UID<string>;
	value(): string;
	isNew(): boolean;
	createdAt(): Date;
	isShort(): boolean;
	equal(id: UID<string>): boolean;
	deepEqual(id: UID<string>): boolean;
	cloneAsNew(): UID<string>;
	clone(): UID<T>;
}

/**
 * 
 */
export interface ITeratorConfig<T> {
	initialData?: Array<T>;
	returnCurrentOnReversion?: boolean;
	restartOnFinish?: boolean;
}

/**
 * 
 */
export interface IIterator<T> {
	hasNext(): boolean;
	hasPrev(): boolean;
	next(): T;
	prev(): T;
	first(): T;
	last(): T;
	isEmpty(): boolean;
	toFirst(): IIterator<T>;
	toLast(): IIterator<T>;
	toArray(): Array<T>;
	clear(): IIterator<T>;
	addToEnd(data: T): IIterator<T>;
	add(data: T): IIterator<T>;
	addToStart(data: T): IIterator<T>;
	removeLast(): IIterator<T>;
	removeFirst(): IIterator<T>;
	total(): number;
	clone(): IIterator<T>;
	removeItem(item: T): void;
}

/**
 * 
 */
export type IResultOptions = 'fail' | 'Ok';

/**
 * 
 */
export interface ICommand<A, B> {
	execute(data: A): B;
}

/**
 * 
 */
export type IUseCase<T, D> = ICommand<T, Promise<D>>;

export interface IProxy<T, D> {
	canExecute:<A extends T>(data: A) => Promise<boolean> | boolean;
	beforeExecute?:<A extends T, B extends T>(data: A) => Promise<B>;
	execute: ICommand<T, D>;
	afterExecute?:<A extends D, B extends D>(data: A) => Promise<B>;
}

export interface ISettings {
	disableGetters?: boolean;
	disableSetters?: boolean;
}

export interface IResultObject<T, D, M> {
	isOk: boolean;
	isFail: boolean;
	data: T | null;
	error: D | null;
	metaData: M;
}

export interface IResultHook<Y> {
	on(option: IResultOptions): Y | undefined;
}

export interface IResultExecute<X, Y> extends IResultHook<Y> {
	withData(data: X): IResultHook<Y>;
}

export type OBJ = { };

export type EntityProps = OBJ | { id?: string, createdAt?: Date, updatedAt?: Date };

export interface EntityMapperPayload {
	id: string,
	createdAt: Date,
	updatedAt: Date
};

export interface IHistoryProps<Props> {
	props: Props;
	action: 'create' | 'update';
	token?: UID<string>;
	ocurredAt?: Date;
}

export interface IHistory<Props> {
	snapshot(props: IHistoryProps<Props>): IHistoryProps<Props>;
	back(token?: UID<string>): IHistoryProps<Props>;
	forward(token?: UID<string>): IHistoryProps<Props>;
	count(): number;
	list(): Array<IHistoryProps<Props>>;
}

export interface IEntityHistory<Props> {
	back(token?: UID<string>): IHistoryProps<Props> | null;
	forward(token?: UID<string>): IHistoryProps<Props> | null;
	count(): number;
	list(): Array<IHistoryProps<Props>>;
}

export interface IPublicHistory<Props> {
	snapshot(token?: UID<string>): IHistoryProps<Props>;
	back(token?: UID<string>): IHistoryProps<Props>;
	forward(token?: UID<string>): IHistoryProps<Props>;
	count(): number;
	list(): Array<IHistoryProps<Props>>;
}

export type IPropsValidation<T> = { [P in keyof Required<T>]: (value: T[P]) => boolean };

/**
 * @description Domain Events Params
 * @param aggregate the entity to add the event.
 * @param createdAt the current date time the event was created.
 * @param callback the event handler to be executed on dispatch.
 */
export interface IDomainEvent<T> {
	aggregate: T;
	createdAt: Date;
	callback: IHandle<T>;
}

/**
 * @description Define a handler to be executed on dispatch an event.
 * @var eventName is optional value as string or undefine.
 * @method dispatch is the method to be executed on dispatch the event.
 */
export interface IHandle<T> {
	/**
	 * @description eventName is optional value. Default is className
	 */
	eventName?: string;
	dispatch(event: IDomainEvent<T>): Promise<void> | void;
}

/**
 * @description Options to dispatch some event.
 * @param eventName is the value defined on handler. Default is the class name.
 * @param id is the ID to identify the aggregate on state.
 */
export interface IDispatchOptions {
	eventName: string;
	id: UID<string>;
}

/**
 * @description Event options
 */
export interface IEvent<G> {
	event: IDomainEvent<G>;
	replace?: boolean;
}

export type IReplaceOptions = 'REPLACE_DUPLICATED' | 'UPDATE' | 'KEEP';

export interface IAdapter<F, T>{
	build(target: F): IResult<T>;
}

export interface IEntity<Props>{
	toObject<T>(adapter?: IAdapter<IEntity<Props>, any>): T extends {} ? T & EntityMapperPayload : { [key in keyof Props]: any } & EntityMapperPayload;
	get id(): UID<string>;
	hashCode(): UID<string>;
	isNew(): boolean;
	clone(): IResult<IEntity<Props>>;
}

export interface IValueObject<Props>{
	clone(): IResult<IValueObject<Props>>;
	toObject<T>(adapter?: IAdapter<this, T>): T;
}



export interface IGettersAndSetters<Props> {
	validation<Key extends keyof Props>(value: Props[Key], key: Key): boolean;
	get<Key extends keyof Props>(key: Key): Props[Key];
	set<Key extends keyof Props>(key: Key): {
		to: (value: Props[Key], validation?: (value: Props[Key]) => boolean) => IGettersAndSetters<Props>
	};
	change<Key extends keyof Props>(key: Key, value: Props[Key], validation?: (value: Props[Key]) => boolean): IGettersAndSetters<Props>;
	history(): IPublicHistory<Props>;
}

export interface IAggregate<Props>{
	toObject<T>(adapter?: IAdapter<this, T>): T extends {} ? T & EntityMapperPayload : { [key in keyof Props]: any } & EntityMapperPayload;
	get id(): UID<string>;
	hashCode(): UID<string>;
	isNew(): boolean;
	clone(): IResult<IEntity<Props>>;
	addEvent(event: IHandle<IAggregate<Props>>, replace?: IReplaceOptions): void;
	deleteEvent(eventName: string): void;
}

export type IParentName = 'ValueObject' | 'Entity';


export interface IAutoMapper<Props> {
	valueObjectToObj(valueObject: IValueObject<Props>): { [key in keyof Props]: any };
	entityToObj(entity: IEntity<Props>): { [key in keyof Props]: any } & EntityMapperPayload;
}

export interface IManyData {
	class: any;
	props: OBJ;
}

export type ICreateManyDomain = Array<IManyData>;

export interface ICreateManyResult {
	data: IIterator<IResult<any, any, any>>;
	result: IResult<any, any, any>;
}

export type IClass = {};

