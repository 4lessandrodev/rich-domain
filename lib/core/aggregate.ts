import { EntityProps, IAggregate, IHandle, IReplaceOptions, IResult, ISettings, UID } from "../types";
import DomainEvent from "./domain-event";
import Entity from "./entity";
import DomainEvents from "./events";
import ID from "./id";
import Result from "./result";

/**
 * @description Aggregate identified by an id
 */
 export class Aggregate<Props extends EntityProps> extends Entity<Props> implements IAggregate<Props> {

	constructor(props: Props, config?: ISettings) { 
		super(props, config);
	}

	/**
	 * @description Get hash to identify the aggregate.
	 * @returns Aggregate hash as ID instance.
	 * @example 
	 * `[Aggregate@ClassName]:UUID`
	 * 
	 * @summary className is defined on constructor config param
	 */
	public hashCode(): UID<string> {
		const name = Reflect.getPrototypeOf(this);
		return ID.create(`[Aggregate@${name?.constructor.name}]:${this.id.value()}`);
	}

	/**
	 * @description Add event to aggregate
	 * @param event Event to be dispatched
	 * @param replace 'REPLACE_DUPLICATED' option to remove old event with the same name and id
	 */
	addEvent(event: IHandle<Aggregate<Props>>, replace?: IReplaceOptions): void {
		DomainEvents.addEvent({ event: new DomainEvent(this, event), replace: replace === 'REPLACE_DUPLICATED' })
	}

	deleteEvent(eventName: string): void {
		DomainEvents.deleteEvent({ eventName, id: this.id });
	}

	/**
	 * 
	 * @param props params as Props
	 * @param id optional uuid as string, second arg. If not provided a new one will be generated.
	 * @returns instance of result with a new Aggregate on state if success.
	 * @summary result state will be `null` case failure.
	 */
	public static create(props: any): IResult<Aggregate<any>, any, any> {
		if(!this.isValidProps(props)) return Result.fail('Invalid props to create an instance of ' + this.name);
		return Result.Ok(new this(props));
	};
}
export default Aggregate;
