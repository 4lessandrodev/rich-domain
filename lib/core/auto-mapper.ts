import { EntityMapperPayload, IAutoMapper, IEntity, IValueObject } from "../types";
import { Validator } from "../utils";
import ID from "./id";

/**
 * @description Auto Mapper transform a domain resource into object.
 */
 export class AutoMapper<Props> implements IAutoMapper<Props> {
	private validator: Validator = Validator.create();

	/**
	 * @description Transform a value object into a simple value.
	 * @param valueObject as instance.
	 * @returns an object or a value object value.
	 */
	valueObjectToObj(valueObject: IValueObject<Props>): { [key in keyof Props]: any } {
		// internal state
		let props = {} as { [key in keyof Props]: any };

		const isSimpleValue = this.validator.isBoolean(valueObject) ||
		this.validator.isNumber(valueObject) ||
		this.validator.isString(valueObject) ||
		this.validator.isObject(valueObject) ||
		this.validator.isDate(valueObject);

		if (isSimpleValue) return valueObject as { [key in keyof Props]: any };

		const isID = this.validator.isID(valueObject);

		const id: ID<any> = valueObject as unknown as ID<any>;

		if (isID) return id?.value() as any;

		// props
		const voProps = valueObject?.['props'];

		const isSimp = this.validator.isBoolean(voProps) ||
		this.validator.isNumber(voProps) ||
		this.validator.isString(voProps) ||
		this.validator.isDate(voProps);

		if (isSimp) return voProps;

		const keys: Array<keyof Props> = Object.keys(voProps) as Array<keyof Props>;

		const values = keys.map((key) => {
			
			const isVo = this.validator.isValueObject(voProps?.[key]);

			if (isVo) return this.valueObjectToObj(voProps?.[key] as any);

			const isSimpleValue = this.validator.isBoolean(voProps?.[key]) ||
				this.validator.isNumber(voProps?.[key]) ||
				this.validator.isString(voProps?.[key]) ||
				this.validator.isObject(voProps?.[key]) ||
				this.validator.isDate(voProps?.[key]);
			
			if (isSimpleValue) return voProps?.[key];

			const isID = this.validator.isID(voProps?.[key]);

			const id: ID<string> = voProps?.[key] as unknown as ID<string>;

			if (isID) return id.value();

			const isArray = this.validator.isArray(voProps?.[key]);

			if (isArray) {
				let arr: Array<any> = voProps?.[key] as unknown as Array<any>;
				const results: Array<any> = [];

				arr.forEach((data) => {

					const result = this.valueObjectToObj(data);
					results.push(result);

				});

				return results;
			}
			
		});

		const hasUniqueValue = values.length === 1;

		props = {} as { [key in keyof Props]: any };

		if (!hasUniqueValue) {
			values.forEach((value, i) => {
				props = Object.assign({}, { ...props }, { [keys[i]]: value })
			});
		}

		return hasUniqueValue ? values[0] : props as any;
	}

	/**
	 * @description Transform a entity into a simple object.
	 * @param entity instance.
	 * @returns a simple object.
	 */
	entityToObj(entity: IEntity<Props>): { [key in keyof Props]: any } & EntityMapperPayload {
		
		let result = {} as { [key in keyof Props]: any };

		const isEntity = this.validator.isEntity(entity);
		
		const isAggregate = this.validator.isAggregate(entity);

		const props = entity?.['props'] ?? {};

		const isValueObject = this.validator.isValueObject(entity);

		const isSimpleValue = this.validator.isBoolean(entity) ||
			this.validator.isNumber(entity) ||
			this.validator.isString(entity) ||
			this.validator.isObject(entity) ||
			this.validator.isDate(entity);
		
		if (isSimpleValue) return entity as any;

		if (isValueObject) return this.valueObjectToObj(entity as any) as any;

		if (isEntity || isAggregate) {

			const id = entity?.id?.value();

			const createdAt = entity['props']['createdAt'];

			const updatedAt = entity['props']['updatedAt'];

			result = Object.assign({}, { ...result }, { id, createdAt, updatedAt  });

			const keys: Array<keyof Props> = Object.keys(props) as Array<keyof Props>;
			
			keys.forEach((key) => {
				
				const isArray = this.validator.isArray(props?.[key as any]);

				if (isArray) {
					const arr: Array<any> = props?.[key as any] as unknown as Array<any> ?? [];

					const subProps = arr.map(
						(item) => this.entityToObj(item as any)
					);

					result = Object.assign({}, { ...result }, { [key]: subProps });
				}

				const isSimple = this.validator.isValueObject(props?.[key as any]) ||
				this.validator.isBoolean(props?.[key as any]) ||
				this.validator.isNumber(props?.[key as any]) ||
				this.validator.isString(props?.[key as any]) ||
				this.validator.isDate(props?.[key as any]);

				if (isSimple) {
					const data = this.valueObjectToObj(props[key as any] as any);

					result = Object.assign({}, { ...result }, { [key]: data });
				}
			});
		}

		return result as any;
	}
}

export default AutoMapper;
