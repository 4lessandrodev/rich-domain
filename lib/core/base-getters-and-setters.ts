import { ICreateManyDomain, IBaseGettersAndSetters, ISettings, UID, IVoSettings } from "../types";
import { BuiltIns } from "../types-util";
import util, { Utils } from "../utils/util";
import validator, { Validator } from "../utils/validator";
import createManyDomainInstances from "./create-many-domain-instance";

/**
 * @description defines getter and setter to all domain instances.
 */
export class BaseGettersAndSetters<Props> implements IBaseGettersAndSetters<Props> {
    protected validator: Validator = validator;
    protected static validator: Validator = validator;
    protected util: Utils = util;
    protected static util: Utils = util;

    protected config: ISettings = { disableGetters: false, disableSetters: false };

    constructor(protected props: Props, config?: IVoSettings) {
        BaseGettersAndSetters.validator = validator;
        BaseGettersAndSetters.util = util;
        this.validator = validator;
        this.util = util;
        this.config.disableGetters = !!config?.disableGetters;
    }

    /**
     * @description Create many domain instances
     * @param data Array of options
     * @returns data and result.
     * @summary result: final result of validating each instance
     * @summary data: all created instances as iterator of Result.
     * @callback Class you can use this function to create the args and define types to Props.
     * 
     * @example
     * 
     * const { result, data } = ValueObject.createMany([
     *   Class<AgeProps>(Age, props),
     *   Class<NameProps>(Name, props),
     *   Class<PriceProps>(Price, props)
     * ]);
     * 
     * result.isOk() // true
     * 
     * const age = data.next() as IResult<Age>;
     * const name = data.next() as IResult<Name>;
     * const price = data.next() as IResult<Price>;
     * 
     * age.value().get('value') // 21
     * 
     */
    public static createMany(data: ICreateManyDomain) {
        return createManyDomainInstances(data);
    }

    /**
     * 
     * @param key the property key you want to get
     * @returns the value of property
     */
    get<Key extends keyof Props>(
        key: Props extends BuiltIns ?
            'value' :
            Props extends Symbol ?
            'value' :
            Props extends any[] ?
            'value' :
            Key
    ): Props extends BuiltIns ?
        Props :
        Props extends Symbol ?
        string :
        Props extends any[] ?
        Readonly<Props> :
        Props extends {} ?
        Readonly<Props[Key]> : Props {
        if (this.config.disableGetters) {
            const instance = Reflect.getPrototypeOf(this);
            throw new Error(`Trying to get key: "${String(key)}" but the getters are deactivated on ${instance?.constructor.name}`);
        };
        if (this.validator.isDate(this.props) || this.validator.isArray(this.props)) {
            return this.props as any;
        }
        if (key === 'value') {
            const isSimpleValue = this.validator.isBoolean(this.props) ||
                this.validator.isNumber(this.props) ||
                this.validator.isString(this.props)

            if (this.validator.isSymbol(this.props)) return (this.props as Symbol).description as any;

            if (isSimpleValue) return this.props as any;

            const isID = this.validator.isID(this.props);

            if (isID) return (this.props as unknown as UID)?.value() as any;
        }
        return this.props?.[key as Key] ?? null as any;
    }

    getRaw(): Readonly<Props> {
        if (typeof this.props !== 'object' || !Array.isArray(this.props) || this.props instanceof Date) {
            return this.props;
        }
        return Object.freeze(this.props);
    }

}

export default BaseGettersAndSetters;
