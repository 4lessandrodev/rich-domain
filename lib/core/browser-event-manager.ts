import { Event, EventManager, EventType } from "../types";
import validateContextEventName from "../utils/validate-context-event-name.util";

export default class BrowserEventManager implements EventManager {
    private events: EventType[];
    private static _instance: BrowserEventManager;

    private constructor(private readonly _window: Window & typeof globalThis) {
        this.events = [];
        if (typeof this._window === 'undefined') {
            throw new Error('BrowserEventManager is not supported');
        }
    }

    static instance(window: Window & typeof globalThis): BrowserEventManager {
        if (BrowserEventManager._instance) return BrowserEventManager._instance;
        BrowserEventManager._instance = new BrowserEventManager(window);
        return BrowserEventManager._instance;
    }

    private getEvent(eventName: string): EventType | null {
        const event = this.events.find((event): boolean => event.eventName === eventName);
        if (event) return event;
        return null;
    }

    subscribe(eventName: string, fn: (event: Event) => void | Promise<void>): void {
        validateContextEventName(eventName);
        if (this.exists(eventName)) return;
        this.events.push({ eventName, callback: fn });
        this._window.sessionStorage.setItem('rich-domain-event:' + eventName, Date.now().toString());
        this._window.addEventListener(eventName, fn as unknown as () => {});
        this._window.addEventListener('beforeunload', (): void => {
            this._window.sessionStorage.removeItem('rich-domain-event:' + eventName);
        });
    }

    exists(eventName: string): boolean {
        const existsOnWindow = !!this._window.sessionStorage.getItem('rich-domain-event:' + eventName);
        const existsInternal = !!this.events.find((evt): boolean => evt.eventName === eventName);
        return existsOnWindow || existsInternal;
    }

    removerEvent(eventName: string): boolean {
        this._window.sessionStorage.removeItem('rich-domain-event:' + eventName);
        const event = this.getEvent(eventName);
        if (!event) return false;
        this.events = this.events.filter((event): boolean => event.eventName !== eventName);
        this._window.removeEventListener(eventName, event.callback);
        return true;
    }

    dispatchEvent(eventName: string, ...args: any[]): void {
        validateContextEventName(eventName);
        if (eventName.includes('*')) {
            const regex = new RegExp(eventName.replace('*', '.*'));
            let i = 0;
            while (this.events[i]) {
                const localEventName = this.events[i].eventName;
                const match = regex.test(localEventName);
                if (match) {
                    this._window.dispatchEvent(new this._window.CustomEvent(localEventName, {
                        bubbles: true,
                        detail: args || []
                    }));
                }
                i++;
            }
            return;
        }

        this._window.dispatchEvent(new this._window.CustomEvent(eventName, {
            bubbles: true,
            detail: args || []
        }));
    }
}
