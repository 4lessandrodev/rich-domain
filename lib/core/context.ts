import platform from "../utils/platform.util";
import BrowserEventManager from "./browser-event-manager";
import { EventManager } from "../types";
import ServerEventManager from "./server-event-manager";

export abstract class Context {
    private static eventManager: EventManager;
    static events(): EventManager {
        if (Context.eventManager) return Context.eventManager;
        if (platform.isNodeJs(global?.process)) {
            Context.eventManager = ServerEventManager.instance();
        } else if (platform.isBrowser(globalThis?.window)) {
            Context.eventManager = BrowserEventManager.instance(globalThis.window);
        } else {
            throw new Error('Could not define platform');
        }
        return Context.eventManager;
    }
}

export default Context;
