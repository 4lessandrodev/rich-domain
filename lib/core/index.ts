export * from './aggregate';
export * from './auto-mapper';
export * from './domain-event';
export * from './entity';
export * from './events';
export * from './getters-and-setters';
export * from './history';
export * from './id';
export * from './iterator';
export * from './result';
export * from './value-object';
export * from './create-many-domain-instance';
