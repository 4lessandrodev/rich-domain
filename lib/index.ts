export * from './core/index';
export * from './utils/validator';
export * from './types';
