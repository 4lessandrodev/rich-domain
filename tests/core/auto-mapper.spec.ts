import { AutoMapper, Entity, ID, Iterator, ValueObject } from "../../lib/core";
import { UID } from "../../lib/types";

describe('auto-mapper', () => {

	describe('value-object', () => {

		it('should convert value to a simple string', () => {

			class StringVo extends ValueObject<{ value: string }> {
				private constructor(props: { value: string }) {
					super(props);
				}
			}

			const vo = StringVo.create({ value: 'hello' });

			const autoMapper = new AutoMapper();

			const result = autoMapper.valueObjectToObj(vo.value());
			
			expect(result).toBe('hello');

		});

		it('should convert value to an object if result has more than one key', () => {

			class StringVo extends ValueObject<{ value: string, age: number }> {
				private constructor(props: { value: string, age: number }) {
					super(props);
				}
			}

			const vo = StringVo.create({ value: 'hello', age: 21 });

			const autoMapper = new AutoMapper();

			const result = autoMapper.valueObjectToObj(vo.value());
			
			expect(result).toEqual({ value: 'hello', age: 21 });

		});

		it('should get boolean with success', () => {

			class StringVo extends ValueObject<{ value: string, isActive: boolean }> {
				private constructor(props: { value: string, isActive: boolean }) {
					super(props);
				}
			}

			const vo1 = StringVo.create({ value: 'hello', isActive: true });
			const vo2 = StringVo.create({ value: 'hello', isActive: false });

			const autoMapper = new AutoMapper();

			const result1 = autoMapper.valueObjectToObj(vo1.value());
			const result2 = autoMapper.valueObjectToObj(vo2.value());
			
			expect(result1).toEqual({ value: 'hello', isActive: true });
			expect(result2).toEqual({ value: 'hello', isActive: false });

		});

		it('should convert array and value to a simple object', () => {

			class StringVo extends ValueObject<{ value: string, notes: number[] }> {
				private constructor(props: { value: string, notes: number[] }) {
					super(props);
				}
			}

			const vo = StringVo.create({ value: 'hello', notes: [1,2,3,4,5,6,7] });

			const autoMapper = new AutoMapper();

			const result = autoMapper.valueObjectToObj(vo.value());
			
			expect(result).toEqual({ value: 'hello', notes: [1,2,3,4,5,6,7] });

		});

		it('should get array from value object', () => {

			class StringVo extends ValueObject<{ value: string }> {
				private constructor(props: { value: string }) {
					super(props);
				}
			}

			const vo = StringVo.create({ value: [1,2,3,4,5,6,7] });

			const autoMapper = new AutoMapper();

			const result = autoMapper.valueObjectToObj(vo.value());
			
			expect(result).toEqual([1,2,3,4,5,6,7]);

		});

		it('should get id value from value object', () => {

			class StringVo extends ValueObject<{ value: ID<string> }> {
				private constructor(props: { value: ID<string> }) {
					super(props);
				}
			}

			const vo = StringVo.create({ value: ID.create('3c5738cf-825e-48b7-884d-927be849b0b6') });

			const autoMapper = new AutoMapper();

			const result = autoMapper.valueObjectToObj(vo.value());
			
			expect(result).toBe('3c5738cf-825e-48b7-884d-927be849b0b6');

		});

		it('should get ids value from value object', () => {

			class StringVo extends ValueObject<{ value: ID<string>[] }> {
				private constructor(props: { value: ID<string>[] }) {
					super(props);
				}
			}

			const ids = Iterator.create({ initialData: ['927be849b0b1', '927be849b0b2', '927be849b0b3'] });

			const IDS: UID<string>[] = [];

			while (ids.hasNext()) {
				IDS.push(ID.create(ids.next()));
			}
			
			const vo = StringVo.create({ value:  IDS });

			const autoMapper = new AutoMapper();

			const result = autoMapper.valueObjectToObj(vo.value());
			
			expect(result).toEqual(["927be849b0b1", "927be849b0b2", "927be849b0b3"]);

		});

		it('should get value object from value object', () => {

			class StringVo2 extends ValueObject<{ value: string, age: number }> {
				private constructor(props: { value: string, age: number }) {
					super(props);
				}
			}

			class StringVo extends ValueObject<{ value: StringVo2, message: string }> {
				private constructor(props: { value: StringVo2, message: string }) {
					super(props);
				}
			}

			const vo = StringVo.create({
				value: StringVo2.create({ value: 'hello', age: 21 }).value(),
				message: 'text'
			});

			const autoMapper = new AutoMapper<{ value: StringVo2, message: string }>();

			const result = autoMapper.valueObjectToObj(vo.value());
			
			expect(result).toEqual({ message: "text", value: { value: 'hello', age: 21 } });

		});

		it('should get date from object in other value object', () => {
			process.env.TZ = 'UTC';
			class StringVo2 extends ValueObject<{ value: Date, age: number }> {
				private constructor(props: { value: Date, age: number }) {
					super(props);
				}
			}

			class StringVo extends ValueObject<{ value: StringVo2, message: string }> {
				private constructor(props: { value: StringVo2, message: string }) {
					super(props);
				}
			}

			const vo = StringVo.create({
				value: StringVo2.create({
					value: new Date('2022-01-01T03:00:00.000Z'),
					age: 21
				}).value(),
				message: 'text'
			});

			const autoMapper = new AutoMapper<{ value: StringVo2, message: string }>();

			const result = autoMapper.valueObjectToObj(vo.value());
			
			expect(result).toEqual({
				message: "text",
				value: {
					value: new Date('2022-01-01T03:00:00.000Z'),
					age: 21
				}
			});

		});

	});

	describe('entity', () => {

		it('should get object from entity', () => {

			class NameVo extends ValueObject<{ value: string }>{
				private constructor(props: { value: string }) {
					super(props)
				}
			}

			class AgeVo extends ValueObject<{ value: number }>{
				private constructor(props: { value: number }) {
					super(props)
				}
			}

			interface Props {
				id?: string;
				name: NameVo;
				age: AgeVo;
				notes: Array<number>;
			}

			class SimpleEntity extends Entity<Props>{
				private constructor(props: Props, config?: any) {
					super(props, config);
				}
			}

			const age = AgeVo.create({ value: 21 }).value();
			const name = NameVo.create({ value: 'some value' }).value();
			const agg = SimpleEntity.create({ id: "1519cb69-9904-4f2b-84e1-e6e95431cf24", age, name, notes: [1,2,3] });

			const user = agg.value();

			expect(user.id.value()).toBe('1519cb69-9904-4f2b-84e1-e6e95431cf24');

			const autoMapper = new AutoMapper<Props>();

			const obj = autoMapper.entityToObj(user);
			
			expect(obj.id).toBe('1519cb69-9904-4f2b-84e1-e6e95431cf24');
			expect(obj.age).toBe(21);
			expect(obj.name).toBe('some value');
			expect(obj.createdAt).toBeInstanceOf(Date);
			expect(obj.updatedAt).toBeInstanceOf(Date);
			expect(obj.notes).toEqual([1,2,3]);

		});

	});

});
