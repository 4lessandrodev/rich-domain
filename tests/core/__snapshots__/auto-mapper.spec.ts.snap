// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auto-mapper complex entity should convert an aggregate to simple object with success 1`] = `
Object {
  "amount": 42,
  "createdAt": 2022-11-27T22:39:58.897Z,
  "detail": "detail info",
  "id": "f25a30cb-294c-4269-8e8c-060403f3a971",
  "item": "some-item",
  "lastSales": Array [
    "some-item",
    "some-item",
    "some-item",
  ],
  "name": "jane",
  "options": Array [
    "a",
    "b",
    "c",
  ],
  "price": 20,
  "product": Object {
    "amount": 42,
    "createdAt": 2022-11-27T22:39:58.897Z,
    "detail": "detail info",
    "id": "f25a30cb-294c-4269-8e8c-060403f3a971",
    "item": "some-item",
    "lastSales": Array [
      "some-item",
      "some-item",
      "some-item",
    ],
    "name": "jane",
    "options": Array [
      "a",
      "b",
      "c",
    ],
    "price": 20,
    "updatedAt": 2022-11-27T22:39:58.897Z,
  },
  "updatedAt": 2022-11-27T22:39:58.897Z,
}
`;

exports[`auto-mapper complex entity should convert an entity to simple object with success 1`] = `
Object {
  "amount": 42,
  "createdAt": 2022-11-27T22:39:58.897Z,
  "detail": "detail info",
  "id": "f25a30cb-294c-4269-8e8c-060403f3a971",
  "item": "some-item",
  "lastSales": Array [
    "some-item",
    "some-item",
    "some-item",
  ],
  "name": "jane",
  "options": Array [
    "a",
    "b",
    "c",
  ],
  "price": 20,
  "updatedAt": 2022-11-27T22:39:58.897Z,
}
`;
